### ISSUE TITLE
Uncaught TypeError in RouteCollection.php:70  when using $container->delegate(new ReflectionContainer());

### ISSUE THREAD

**Issue by danhofmann-com** [Xocotlah/Router#1]
So your package seems to break if you use 
 ```php
use League\Container\ReflectionContainer;
use League\Container\Container;

$container = new Container();
$container->delegate(new ReflectionContainer());
 ```

The culprit seems to be $container->delegate(new ReflectionContainer());

I installed the service provider and created the router as follows:
 
 ```php
use League\Container\ReflectionContainer;
use League\Container\Container;

$container = new Container();
$container->delegate(new ReflectionContainer());
 
$container->addServiceProvider( Xocotlah\Router\ServiceProvider\RouterServiceProvider::class);
 
…
 
$router = $container->get('router');
 
…
 
$router->get('/', 'App\Http\Controllers\HomepageController::index');
 
…
 
$response = $router->dispatch(
    $container->get('request'),
    $container->get('response')
);

$container->get('emitter')->emit($response);
 ```
 
But then I get the error below:

```php
Fatal error: Uncaught TypeError: Argument 1 passed to League\Route\RouteCollection::__construct() must implement interface Psr\Container\ContainerInterface, string given in vendor/league/route/src/RouteCollection.php:70 Stack trace: #0 [internal function]: League\Route\RouteCollection->__construct('Psr\\Container\\C...', 'FastRoute\\Route...', 'FastRoute\\DataG...') #1 vendor/league/container/src/ReflectionContainer.php(38): ReflectionClass->newInstanceArgs(Array) #2 vendor/league/container/src/Container.php(260): League\Container\ReflectionContainer->get('Xocotlah\\Router...', Array) #3 vendor/league/container/src/Container.php(93): League\Container\Container->getFromDelegate('Xocotlah\\Router...', Array) #4 vendor/l in vendor/league/route/src/RouteCollection.php on line 70
```

 when my controller extends your BaseController.

 
 ```php
use Xocotlah\Router\Controller\BaseController as XocotlahBaseController;
 
…
 
class HomepageController extends XocotlahBaseController
 
…
 ``` 
 
But, if I remove the following line from my League container implementation:

$container->delegate(new ReflectionContainer());

I no longer get any errors.

However, removing that line causes inflectors to stop working (killing eager loading), which is a show stopper for me.

Is there anyway to get your package to work with 
$container->delegate(new ReflectionContainer()); ?

My code for the container is:

```php
use League\Container\ReflectionContainer;
use League\Container\Container;

$container = new Container();
$container->delegate(new ReflectionContainer());
$container->addServiceProvider( Xocotlah\Router\ServiceProvider\RouterServiceProvider::class);

...
```

I am using the following in my composer.json:
 
"require": {
    
    "zendframework/zend-diactoros": "^1.6",
    "league/route": "^3.0",
    "league/container": "^2.4",
    "xocotlah/router": "^0.1.0"
},

**Comment by Arcesilas** [comment]
So, I've reproduced the bug. Indeed, when disabling autowiring on League/Container, it works.

Just a note about autowiring and inflectors: autowiring is not required in order to use inflectors. Inflectors are always called before returning the instance, they are not related to ReflectionContainer.

Anyway, the router should work when autowiring is activated.

Thanks for spotting this! I'll need to update my tests too...

**Comment by Arcesilas** [comment]
Ok. So, I've considered when writing this lilbrary, that one may always use the controller default namespace.

Try this:

```php
$container->add('controller.prefix', 'App\\Http\\Controllers\\');
$container->addServiceProvider(Xocotlah\Router\ServiceProvider\RouterServiceProvider::class);
```

Then, define the route:
```php
$router->get('/', ['HomepageController','index']);
```

This should work, because the container does not find a callable named HomepageController. Then, the Factory is called to create the appropriate controller (See https://github.com/Xocotlah/Router/blob/master/src/Route.php#L30).

Autowiring makes i possible for the container to instanciate the controller when you pass the Fully Qualified Class Name. Then, the controller factory is not used, and it fails.

Thanks again for spotting this, I'll try to provide a fix in the next few days (after I've moved to my new place...)

**Comment by danhofmann-com** [comment]
It's weird, because i could not get my inflectors to work (to call my setter method on AwareTrait when the trait extends an Interface, that is registered with the inflector...)

 It just returns null...

For example, if I have the following code:


This sets the inflector on ConfigAwareInterface, for which I want the setConfig method to be called and inject the config from the ConfigServiceProvider whenever a class implements the ConfigAwareInterface...

```php
$container = new Container();
$container->addServiceProvider(App\ServiceProvider\ConfigServiceProvider::class);
$container->inflector(\App\Contract\ConfigAwareInterface::class)
    ->invokeMethod('setConfig', ['config']);

```
This is my service provider...

```php
use League\Container\ServiceProvider\AbstractServiceProvider;

class ConfigServiceProvider extends AbstractServiceProvider
{
    /**
     * @var array
     */
    protected $provides = ['config'];

    /**
     * @{inheritDoc}
     */
    public function register()
    {
     $this->getContainer()->share('config', function() {
        return include __DIR__ . '/../../config/app.php';
     });
    }
}
```

The interface that I want to be used for the inflector

```php
interface ConfigAwareInterface
{
    public function setConfig(array $config);

    public function getConfig();

}

``

And finally my trait, that I want to use in my controller:

```php
trait ConfigAwareTrait
{
    public $config;

    public function setConfig(array $config) {
        $this->config = $config;
    }

    public function getConfig() {
        return $this->config;
    }
}

```

Then my BaseContoller implements the ConfigAwareInterface interface

```php

class BaseController extends Controller implements ConfigAwareInterface
{
    use ConfigAwareTrait;
 

```


I noticed that the setConfig() method of the ConfigAwareTrait does not get called. When I call $this->getConfig()  in my controller action, it returns null. I was very puzzled by this. Finally, looking at the hello world 3.0 documentation for reflectors, I noticed I was missing one line:

```php
$container->delegate(new ReflectionContainer());
```

When i added this, and hit refresh on browser.. boom -- my debug statement in the setConfig in the trait fired, and $this->config was no longer null.

However, if I remove the delegate ReflectionContainer line, it no longer works...

Am I doing something wrong here...

**Comment by danhofmann-com** [comment]
Ahh.. I read your last post too hastily.. I see what you mean by the autowiring causing the factory method to be skipped... I can set the default namespace and remove the fully qualified class path to get around this error for now. Thanks for the quick fix on that. I'll try that and await your patch.

**Comment by Arcesilas** [comment]
I'm pretty sure the inflectors don't need autowiring. They are applied in the [get() method](https://github.com/thephpleague/container/blob/4f00d3b7817ee25b9eec52cabb56cd18b1a98afd/src/Container.php#L82) and in the  protected [getFromThisContainer() method](https://github.com/thephpleague/container/blob/4f00d3b7817ee25b9eec52cabb56cd18b1a98afd/src/Container.php#L279).

Anyway, I think I've found a way to fix the bug you reported (while having breakfast). I should consider asking the factory to try first then to ask the container if it has an appropriate instance.

**Comment by danhofmann-com** [comment]
I believe you about the inflectors not needing autowiring, but they definitely do not work for me with autowiring disabled. If you see any errors in the code I pasted, let me know.

 

Over breakfast, LOL, yeah, usually solve problems when I am not in front of the computer either. It’s the developer way… HAHA.

**Comment by Arcesilas** [comment]
Fixed in v0.1.1.
Can you please update the package with Composer and let me know if it's ok for you?
